### PR TITLE
Memory optimization of static_variant

### PR DESCRIPTION
**Pull request by crypto-ape**
In [bitshares/bitshares-core](https://github.com/bitshares/bitshares-core) `operation` is typedef of `static_variant`. As blockchain operations are used everywhere in the database it would be great to reduce the memory footprint of this structure as much as possible.

`static_variant` contains member variable `storage` that is a static array of characters which size is deduced from the largest type of a variant. This means every `operation` in the database is as large as the largest type + 8 bytes for `_tag` + some bytes for memory alignment.

The largest operation is the `account_create_operation` with 424 bytes while the average size of a operation is 110 bytes. *And this operation isn't even a frequently used one*.

I have decided to replace this static array with dynamically allocated storage (`dynamic_storage`).

Here are the results of the optimization:

## Parameters:
*Total block count*: `28387605` 
Important lines in my `config.ini`:
```ini
seed-nodes = []
# partial-operations = 1 // this means partial-operations is 0 (false)
max-ops-per-account = 100
bucket-size = [60,300,900,1800,3600,14400,86400]
history-per-size = 1000
max-order-his-records-per-market = 1000
max-order-his-seconds-per-market = 259200
```
*Input parameters*:
```shell
witness_node --replay-blockchain
```

## Results:

* Original implementation (static allocation):

memory:
```
 Private  +   Shared  =  RAM used       Program
122.0 GiB + 794.0 KiB = 122.0 GiB       witness_node
```
time of the blockchain replay & db reindexing:
```
CPU-seconds in user mode: 5115.06 
CPU-seconds in kernel mode: 142.79
real time: 1:32:42
CPU time: 94%CPU
```

* Optimized (using `std::vector`):

memory:
```
 Private  +   Shared  =  RAM used       Program
 62.5 GiB + 871.0 KiB =  62.5 GiB       witness_node
```

time of the blockchain replay & db reindexing:
```
CPU-seconds in user mode: 5059.77
CPU-seconds in kernel mode: 130.11
real time: 1:30:09
CPU time: 95%CPU 
```

* Optimised (using `raw array`):

memory:
```
 Private  +   Shared  =  RAM used       Program
 55.3 GiB + 573.0 KiB =  55.3 GiB       witness_node
```

time of the blockchain replay & db reindexing:
```
CPU-seconds in user mode: 5058.16
CPU-seconds in kernel mode: 133.52
real time: 1:31:03
CPU time: 95%CPU 
```

## Conclusion

Memory optimization doesn't seem to have effect on the blockchain replay performance.
On the other hand it seems to drastically reduce memory requirements for large amount of operations in the database (in this case almost 67GiB what is around 55%)

*This pull request contains code from the last test, i.e. optimization using `raw array`.*

### CODE REVIEW

File: src/static_variant.cpp
Comment (pmconrad) [inline src/static_variant.cpp]: Please replace with an ``FC_ASSERT``.

**Comment by pmconrad** [comment]
Interesting, thanks!

I think the original design preferred stack allocation over heap allocation for fear of heap management slowing things down. Good to see that this doesn't seem to be the case.

This will conflict with #79 , may have to rebase after merging. @abitmore what's the status of that one?

**Comment by ryanRfox** [comment]
Welcome @crypto-ape to BitShares development. Are you willing to reach out to me via Keybase, Telegram or WeChat? I am `ryanRfox` on those platforms. Hope to discuss further about the Bounty Process and compensation described in the [Core Team worker proposal](https://www.bitshares.foundation/workers/2018-03-bitshares-core).

**Comment by abitmore** [comment]
I've merged #79, so need to rebase this to latest master. Thanks.

**Comment by pmconrad** [comment]
The ``type_in_typelist`` template does its job, however, I find the error message from the ``static_assert`` much more descriptive. Is there a particular reason why you replaced it?

**Comment by abitmore** [comment]
A reason for using a static storage, is it's more computational efficient to update a `static_variant` variable from one type to another type, since it avoided additional memory free-up and reallocation. Specifically, `set_which()` and `operator=()` are affected. Current code free and reallocate memory even when type of contained data is unchanged which IMHO is unnecessary.

If profiling of replay shows no significant penalty when moved to a dynamic storage, it means we're not updating `static_variant` objects in such way frequently during replay.
- [ ] if we profile with full re-validation(see https://github.com/bitshares/bitshares-core/pull/1360), the result may or may not be different
- [ ] IMHO we should check other code that updates `static_variant` objects.
  - [ ] One place I'm aware of is using `set_which()` is the `get_required_fees` API call, which won't be called in replay, but is heavily used by clients. It means performance of API nodes would be affected by this change. The call of `set_which()` there can be avoided IMHO.
  - [ ] `set_fees` is similar
  - [ ] other code

**Comment by crypto-ape** [comment]
> The `type_in_typelist` template does its job, however, I find the error message from the `static_assert` much more descriptive. Is there a particular reason why you replaced it?

Using `type_in_typelist` (SFINAE method) has few advantages:

1. It's reusable, you don't have to copy+paste the same condition & the error message for every templated method. `static_assert` might be reused as well, e.g. with function like `void assert_type_in_typelist<X, T...>()`, except the case of the `struct tag`.

2. It removes a function from *overloaded resolutions* when a condition is false. This means function with incorrect signature won't be even 'created'.
    * You will not have problems with function overloading. What if you would want to have `static_variant` constructed with `integral type`.
        ```cpp
        static_variant(_tag_type tag){
            ...
        }
        ```
        And imagine you would do something like this: `static_variant<a,b,c> var(0)` -> linker would find templated version more suitable because `0` is of type `int` not `_tag_type`.
    * Modern IDEs usually show error already when your are typing function with incorrect parameters. So you don't have to compile your code to see that you have an error in the code.


Disadvantage is, that you can't specify error message and you will get message like this: `template argument deduction/substitution failed: ...` .

In fact I decided to use `SFINAE` instead of `static_assert`, when I encountered problems with overloading during development. I was experimenting with more generic version of variant with `storage` as a template argument.


> A reason for using a static storage, is it's more computational efficient to update a `static_variant` variable from one type to another type, since it avoided additional memory free-up and reallocation. Specifically, `set_which()` and `operator=()` are affected. Current code free and reallocate memory even when type of contained data is unchanged which IMHO is unnecessary.
> 
> If profiling of replay shows no significant penalty when moved to a dynamic storage, it means we're not updating `static_variant` objects in such way frequently during replay.
> 
> * [ ]  if we profile with full re-validation(see [bitshares/bitshares-core#1360](https://github.com/bitshares/bitshares-core/pull/1360)), the result may or may not be different
> * [ ]  IMHO we should check other code that updates `static_variant` objects.
>   
>   * [ ]  One place I'm aware of is using `set_which()` is the `get_required_fees` API call, which won't be called in replay, but is heavily used by clients. It means performance of API nodes would be affected by this change. The call of `set_which()` there can be avoided IMHO.
>   * [ ]  `set_fees` is similar
>   * [ ]  other code

Reallocation of memory might be avoided with one if & some changes in `dynamic_storage`:
```cpp
size_t type_size( tag_type tag )
{
    return impl::size<tag_type, Types...>( tag )
}

void set_which( tag_type w ) 
{
    if ( type_size( _tag ) >= type_size( w ) )
    {
        impl::storage_ops<0, Types...>::del( _tag, storage.data() );
    }
    else
    {
        clean();
    }

    init(w);
}
```
Simillary with `operator=()`.

But this might cause to have garbage in our memory as before.
e.g. this case:
```cpp
operation ac( account_create_operation() )
ac.set_witch(0);
```

Now there would be operation with small memory footprint (`transfer_operation` with `176 bytes`), but it would take up space of the largest operation `account_create_operation` with `424 bytes`.

So having to reallocate storage is trade-off between speed & memory. IMHO reallocation shouldn't be a problem.

What do you mean by this?
> if we profile with full re-validation(see bitshares/bitshares-core#1360), the result may or may not be different

I tested this before any `revalidate-blockchain` flag was there & I thought `replay-blockchain` have to do validation as well.

**Comment by pmconrad** [comment]
> What if you would want to have static_variant constructed with integral type. 

Good point. What if ``Types`` contains ``tag_type`` or ``int64_t``?
Can you add some unit tests?

> I thought replay-blockchain have to do validation as well.

replay skips most validation for better performance, assuming that validation has already happened when the blocks were received for the first time.

**Comment by crypto-ape** [comment]
> > What if you would want to have static_variant constructed with integral type.
> 
> Good point. What if `Types` contains `tag_type` or `int64_t`?
> Can you add some unit tests?

I think you found a bug in my code.
```cpp
    int64_t zero = 2;
    fc::static_variant<int64_t, std::string, int> f(zero);
    printf("which: %d\n", f.which() );
```
This code should return `0`, but returns `2`. Why? Because the overloaded `init(tag_type tag)` function is called in the constructor.
I should add some tests as you suggest.

> > I thought replay-blockchain have to do validation as well.
> 
> replay skips most validation for better performance, assuming that validation has already happened when the blocks were received for the first time.

OK, maybe I'll try to do some test with re-validation.

**Comment by crypto-ape** [comment]
I've created untittest for some special cases of `static_variant` including case when type `T` is `tag_type`.

I've also done some performance tests using the same binaries as before.
## 1) Force revalidation when replaying blockchain

Tests were done on different machine with **fewer blocks**, so you shouldn't compare these results with the previous one. I've run the code 3 times (+3 different implementations) with skip flags = 0 (nothing should be skipped).

*Total block count:* `26033534`

## Results:

* Original implementation (static allocation):

time of the blockchain replay & db reindexing:

```
CPU-seconds in user mode: 3329.09
CPU-seconds in kernel mode: 126.09
real time: 59:36.05
CPU time: 96%CPU 
```
```
CPU-seconds in user mode: 3384.91
CPU-seconds in kernel mode: 122.07
real time: 1:01:39
CPU time: 94%CPU 
```
```
CPU-seconds in user mode: 3531.35
CPU-seconds in kernel mode: 117.96
real time: 1:03:28
CPU time: 95%CPU 
```

* Optimized (using `std::vector`):

time of the blockchain replay & db reindexing:
```
CPU-seconds in user mode: 3338.00
CPU-seconds in kernel mode: 105.92
real time: 59:27.28
CPU time: 96%CPU
```
```
CPU-seconds in user mode: 3366.12
CPU-seconds in kernel mode: 109.52
real time: 59:48.85
CPU time: 96%CPU
```
```
CPU-seconds in user mode: 3348.28
CPU-seconds in kernel mode: 108.16
real time: 1:00:15
CPU time: 95%CPU 
```

* Optimized (using `raw array`):

time of the blockchain replay & db reindexing:
```
CPU-seconds in user mode: 3315.86
CPU-seconds in kernel mode: 104.72
real time: 58:49.67
CPU time: 96%CPU
```
```
CPU-seconds in user mode: 3351.83
CPU-seconds in kernel mode: 103.19
real time: 59:43.50
CPU time: 96%CPU
```
```
CPU-seconds in user mode: 3485.51
CPU-seconds in kernel mode: 108.48
real time: 1:01:43
CPU time: 97%CPU 
```

Again my change has negligible impact on performance.

## 2)

To prove that a reallocation of memory when calling `set_which` isn't such problem, I've run this code:

```cpp
    template<unsigned int N> using arr = std::array<char, N>;
    ...

    fc::static_variant< 
        arr<10>, 
        arr<100>, 
        arr<400>, 
        arr<200>, 
        arr<201>, 
        arr<31>, 
        arr<101>, 
        arr<121>, 
        arr<91>, 
        arr<51>, 
        arr<122>, 
        arr<61>, 
        arr<161>, 
        arr<221>
    > tested;

    using namespace std::chrono;

    double sum = 0;

    const size_t round_count = 1000;
    const size_t it_count = 1000000;

    for (size_t k = 0; k < round_count; k++)
    {
        auto t_start = high_resolution_clock::now();

        for (size_t i = 0; i < it_count; i++)
        {
            for (sv::tag_type j = 0; j < tested.count(); j++)
            {
                tested.set_which(j);

                if (tested.which() != j)
                {
                    BOOST_ASSERT(false);
                }
            }
        }

        auto t_end = high_resolution_clock::now();
        auto elapsed = t_end - t_start;

        sum += duration<double, std::milli>(elapsed).count();
    }

    printf( "Time %f ms\n", sum/(double)round_count );
```

Calling `set_which` totally 1000000 * 14 times took in average `537.406225ms` on my `Intel Core i5-8250U`.
So on an average laptop you can call `set_which` 14 million times in 0.5 seconds. If it isn't called much more often, I don't think it's worth optimization.

**Review by pmconrad — APPROVED** [review]
Excellent, thanks!

**Comment by crypto-ape** [comment]
Hi guys, is there something else I can do to merge this PR?
I don't have write access, so it's up to someone responsible.

@ryanRfox 
> Welcome @crypto-ape to BitShares development. Are you willing to reach out to me via Keybase, Telegram or WeChat? I am `ryanRfox` on those platforms. Hope to discuss further about the Bounty Process and compensation described in the [Core Team worker proposal](https://www.bitshares.foundation/workers/2018-03-bitshares-core).

Sorry for the late response, I will contact you in the near future.

**Comment by ryanRfox** [comment]
@crypto-ape our process for merging requires two (2) Core Devs to review and approve the PR. Thus far we have one, so I'll invite @abitmore or @jmjatlanta to take a look now. Thereafter it will be merged by a team member with write access.

Please do reach out, as I'd like to compensate your efforts. Thanks!

**Comment by pmconrad** [comment]
Thanks!

**Pull request merged**
